### ISSUE TITLE
cif_core.dic: add examples of the Matrix container type values

### ISSUE THREAD

**Issue by vaitkus** [COMCIFS/cif_core#269]
The `CIF_CORE` dictionary defines several data items that have the `Matrix` container type, however, none of these definitions contain value examples (e.g. using the `DESCRIPTION_EXAMPLE` loop).

Some definitions have dREL snippets that explicitly assign Matrix values, but the matrix notation seems to also depend on the dimensions of the item.

For example, data item `_atom_site.fract_xyz` has the dimensions of `[3]` and is assigned as:
```
_atom_site.fract_xyz =  [a.fract_x, a.fract_y, a.fract_z]
```
while data item `_atom_site_aniso.matrix_B` has the dimensions of `[3,3]` and is assigned as:
```
a.matrix_B =  [[ a.B_11, a.B_12, a.B_13 ],
               [ a.B_12, a.B_22, a.B_23 ],
               [ a.B_13, a.B_23, a.B_33 ]]
```
Note, that in the first case the matrix is expressed as a list while in the second example it is expressed as a list of lists.

Having at least a few explicit examples that cover both of the previously mentioned cases would aid greatly in understanding the the way these specific data items as well as the overall Matrix container is intended to be used.

It might also be useful to introduce an `example/` directory with an example CIF file that contains data items of this kind.

**Comment by jamesrhester** [comment]
I agree that some examples in an examples directory would be good. We need to be careful to distinguish between dREL notation and CIF2 notation, e.g. CIF2 does not use commas as separators. We also need to distinguish between a DDLm type and the CIF2 type. So the only CIF2 type is a `List` of strings or other `Lists`, and then DDLm explains how to interpret that list.  I think the `ddl.dic` explanation is pretty good on explaining the meaning of `Matrix`, is there anything missing from it?

**Comment by rowlesmr** [comment]
> Is this OK or would it be better to recommend using a List or an Array container type for such data items?

I'm literally, right now, writing an issue on this in PD (I think they should be `List`s)

**Comment by rowlesmr** [comment]
> 1. Let's say we have a Matrix data item, that has a dimension of `'[4]'`. 

3?

> Is it ok to write in CIF2 as `[ 'a' 'b' 'c' ]` or should it always be double bracketed in some way, i.e. `[[ 'a' 'b' 'c' ]]` or `[[ 'a' ] [ 'b' ] [ 'c' ]]`. The `ddlm_validate` validator currently expects there to always be nested lists so I was wondering if I should change that.

Just somthing from (the way I've been using it in the last month) numpy: `[ 'a' 'b' 'c' ]` is a row vector, and `[[ 'a' ] [ 'b' ] [ 'c' ]]` is a column vector.

.

Can a `Matrix` have more than two dimensions? If so, then double bracketing isn't the way to go, as you would need to account for an arbitrary number of dimensions.


Just for reference:

`_type.dimension` says
>A Matrix with a single dimension is interpreted as a vector

whereas the `Matrix` definitions says
>A matrix with a single dimension is interpreted as a row or column vector as required.

.

**Comment by vaitkus** [comment]
@rowlesmr Thank for catching the typo in the question (4 -> 3).

The interpretation of the `_type.dimension` is more or less clear:
* When the dimension has two values, e.g. `[3, 4]` they correspond to the number of matrix row and columns accordingly. * When the dimension has a single value `[3]`, they correspond to a vector, that can be interpreted as a either be a row vector or a column vector, whichever is more convenient. In other words, they can be interpreted either as `[3,1]` or `[1,3]` matrices.
* Matrices with more than 2 dimensions are not supported.

However, as mentioned by @jamesrhester, CIF2 does not natively have a `Matrix` or a `Vector` data type, only the list syntactical construct (`[ ... ]`) thus leading to some ambiguities in how it should be represented.

Clearly, a 2-D matrix will be composed of multiple nested CIF2 lists (`[ [ 'a' 'b' ] [ 'c' 'd' ] ]`), but it is not clear where a Matrix of a single dimension can/should be expressed in the same way as the List/Array containers (`[ 'a' 'b' ]`) or if matrices should always be written as a list of lists (`[[ 'a' 'b' ]]`).

**Comment by jamesrhester** [comment]
> Clearly, a 2-D matrix will be composed of multiple nested CIF2 lists ([ [ 'a' 'b' ] [ 'c' 'd' ] ]), but it is not clear where a Matrix of a single dimension can/should be expressed in the same way as the List/Array containers ([ 'a' 'b' ]) or if matrices should always be written as a list of lists ([[ 'a' 'b' ]]).

This should be clarified in the description of how CIF syntactical items are mapped to DDLm items in the volume G write-up. I would suggest that either double-bracketed or single-bracketed are both acceptable, as there is no ambiguity.

**Comment by vaitkus** [comment]
I would be really interested in taking a peak at the draft specification once it is ready. We could then prepare the examples based specifically on that draft.

I am still interested though how `Matrix` with dimension `[]` should be interpreted. Matrix of any size? Vector of any length? Or we could simply explicitly disallow such combinations.

**Comment by vaitkus** [comment]
> I would say that if you want to retain the ability to have a single dimension matrix to be interpreted as a column or row vector as necessary, then single bracketing is the way to go.
> 
> With double bracketing (to my mind), '[[1 2 3 4]]' is a row vector of four elements, and '[[1] [2] [3] [4]]' is a column vector of four elements.

We could theoretically have all three:
1. `[1 2 3 4]` -- specifies a vector which can be interpreted either as a column or as a row vector.
2. `[[1 2 3 4]]` -- strictly a row vector.
3. `[[1] [2] [3] [4]]` -- strictly a column vector.

2, 3 would be used only when the distinction is really important (not that I can really think of such a situation). Actually, if the distinction is so important, it would probably also be reflected by the dimensions, i.e. them being set to `[n,1]` or `[1,n]` instead of [`n`].

**Comment by jamesrhester** [comment]
One of the simplifications of dREL is that a vector can act as a row or column vector to suit. The classic example is `_refln.hkl`, which might be both pre- and post-multiplied in a single expression. Cases (2) and (3) can be interpreted as @vaitkus suggests, but they would both be `Matrix` types.

**Comment by rowlesmr** [comment]
> We could theoretically have all three:

I like this idea. There should be some examples to show how it works... ;)
